<p style="color:#153462; 
          font-weight: bold; 
          font-size: 30px; 
          font-family: Gill Sans, sans-serif;
          text-align: center;">
          Implementation of NN in Python</p>

<p style="text-align: justify; 
          text-justify: inter-word;
          font-size:17px;">
    The main goal of this notebook is implementing simple artificial neural network from end to end.
</p>

 ### <span style="color:#C738BD; font-weight: bold;">Simple Manual Calculation of Layer</span>

<img src="images\basic_nn.png" alt="basic_nn" style="width: 500px;"/>

<p style="text-align: justify; 
          text-justify: inter-word;
          font-size:17px;">
  Let's assume you have 4 input nodes and 3 output nodes. Now you have $4 * 3 = 12$ weights and 3 bias(to know more about bias look it in nn_terminology notebook)
</p>

In [2]:
input = [1, 2.3, 3, 0.5]

# Random weights
weight_1 = [0.5, 0.4, 0.9, 0.6]
weight_2 = [0.6, 0.2, 0.1, 0.4]
weight_3 = [0.6, 0.1, 0.5, 0.6]

#Bias
bias_1 = 1
bias_2 = 2
bias_3 = 3

# output of each output node
output = [input[0] * weight_1[0] +  input[1] * weight_1[1] + input[2] * weight_1[2]+ input[3] * weight_1[3] + bias_1,
          input[0] * weight_2[0] +  input[1] * weight_2[1] + input[2] * weight_2[2] + input[3] * weight_2[3] + bias_2,
          input[0] * weight_3[0] +  input[1] * weight_3[1] + input[2] * weight_3[2] + input[3] * weight_3[3] + bias_3]
print(output)

[5.42, 3.56, 5.63]


 ### <span style="color:#C738BD; font-weight: bold;">With For loop</span>

In [1]:
inputs = [1, 2.3, 3, 0.5]

# Random weights
weights = [[0.5, 0.4, 0.9, 0.6], [0.6, 0.2, 0.1, 0.4], [0.6, 0.1, 0.5, 0.6]]
biases = [1, 2, 3]

layer_output = []
for weight, bias in zip(weights, biases):
    neuron_output = 0
    for n_input, w in zip(inputs, weight):
        neuron_output += n_input * w
    neuron_output += bias
    layer_output.append(neuron_output)

print(layer_output)

[5.42, 3.56, 5.63]


 ### <span style="color:#C738BD; font-weight: bold;">Implementation with Numpy</span>

In [2]:
import numpy as np

inputs = [1, 2.3, 3, 0.5]

# Random weights
weights = [[0.5, 0.4, 0.9, 0.6], [0.6, 0.2, 0.1, 0.4], [0.6, 0.1, 0.5, 0.6]]
biases = [1, 2, 3]

# Dot product
# The dimention is very important while performing dot product
# inputs has (4, 1) and weights has (3, 4), so meet dot product criteria either do transportation of weight or
# do weights * inputs
layer_output = np.dot(weights, inputs) + biases
print(layer_output)

[5.42 3.56 5.63]


 ### <span style="color:#C738BD; font-weight: bold;">Batches, Layer and Objects</span>

<p style="text-align: justify; 
          text-justify: inter-word;
          font-size:17px;">
  In this sectioin, we are going to feed a batch of inputs. Batch of inputs helps for better
  generalization. When you feed a batch of inputs, the network performs the same operations 
  (using the same weights) on all inputs simultaneously. This parallel processing is what makes
  batching computationally efficient, especially on GPUs. During training, the network updates its
  weights based on the error calculated from the entire batch.The updates aim to improve performance
  across all inputs in the batch, leading to more stable and efficient training.
</p>


In [5]:
# It has size of 4 X 4
batch_of_inputs = [[1, 2.3, 3, 0.5],
                   [0.5, 1, 3, 0.4],
                   [4.5, 3, 1, 0.9],
                   [0.5, 0.4, 2, 0.2]]

# Random weights
# It has size of 3 X 4
weights = [[0.5, 0.4, 0.9, 0.6], 
           [0.6, 0.2, 0.1, 0.4], 
           [0.6, 0.1, 0.5, 0.6]]

biases = [1, 2, 3]

In [17]:
import numpy as np

# Performing transpose to make it compatible with matrix multiplication
output =  np.dot(batch_of_inputs, np.transpose(weights))
# NOTE: The number of batches of input equals to number of rows of the output
output

array([[4.42, 1.56, 2.63],
       [3.59, 0.96, 2.14],
       [4.89, 3.76, 4.04],
       [2.33, 0.66, 1.46]])

In the output matrix:


\begin{bmatrix}
4.42 & 1.56 & 2.63 \\
3.59 & 0.96 & 2.14 \\
4.89 & 3.76 & 4.04 \\
2.33 & 0.66 & 1.46
\end{bmatrix}


- Each **row** corresponds to an input sample from the `batch_of_inputs`.
- Each **column** corresponds to a neuron in the layer (since there are 3 neurons, we have 3 columns).
- The values in the matrix represent the weighted sum of inputs for each neuron before applying any activation function.

For example:
- The first row `[6.4, 2.9, 4.7]` represents the weighted sum of inputs for each of the 3 neurons for the **first input sample** `[1, 2, 3, 4]`.
- The second row `[3.59, 0.96, 2.14]` represents the weighted sum of inputs for the 3 neurons for the **second input sample** `[0.5, 1, 3, 0.4]`.

Thus, each row in the output represents the pre-activation outputs of all neurons for a single input sample.

In [9]:
# If you see carefully, first row output matches with single input output
output + biases

array([[5.42, 3.56, 5.63],
       [4.59, 2.96, 5.14],
       [5.89, 5.76, 7.04],
       [3.33, 2.66, 4.46]])

#### <span style="color:#40A578; font-weight: bold;">Adding Additional Layer</span>

<p style="text-align: justify; 
          text-justify: inter-word;
          font-size:17px;">
  Adding an additional layer with 3 neurons. The output of second layer going to act as an input now, so we just only need to defined weights and bias.
</p>

In [10]:
# It has size of 4 X 4
batch_of_inputs = [[1, 2.3, 3, 0.5],
                   [0.5, 1, 3, 0.4],
                   [4.5, 3, 1, 0.9],
                   [0.5, 0.4, 2, 0.2]]

# Random weights, It has size of 3 X 4
# Since we have 4 input nodes we are going to have 4 columns and
# we have 3 output nodes so we are going to have 3 rows 
weights1 = [[0.5, 0.4, 0.9, 0.6], 
           [0.6, 0.2, 0.1, 0.4],
           [0.6, 0.1, 0.5, 0.6]]

# Since our input layer has 3 neurons we are going to have 3 columns
# and since we have 3 neurons in the output layer we are going to have 3 rows.
weights2 = [[-0.1, 0.7, -0.3],
            [0.4, -0.2, 0.12],
            [0.3, 0.1, 0.2]]

bias1 = [1, 2, 3]
bias2 = [0.2, 0.5, 0.3]

layer1_output = np.dot(batch_of_inputs, np.array(weights1).T) + bias1
layer2_output = np.dot(layer1_output, np.array(weights2).T) + bias2
print(layer2_output)

[[0.461  2.6316 3.408 ]
 [0.271  2.3608 3.001 ]
 [1.531  2.5488 4.051 ]
 [0.391  1.8352 2.457 ]]


#### <span style="color:#40A578; font-weight: bold;">Converting Into an Object</span>

In [16]:
import numpy as np

np.random.seed(0)

X = [[1, 2.3, 3, 0.5],
     [0.5, 1, 3, 0.4],
     [4.5, 3, 1, 0.9],
     [0.5, 0.4, 2, 0.2]]


class LayerDense:
    def __init__(self, n_inputs, n_neurons):
        """
        :param n_inputs: Input size or num of features, in our example we have 4 features
        :type  n_inputs: int
        :param n_neurons: Number of neurons in the ouput layer
        :type  n_neurons: int
        """
        self.weights = np.random.randn(n_inputs, n_neurons)
        self.bias = np.zeros((1, n_neurons))
    
    def forward(self, inputs):
        self.output = np.dot(inputs, self.weights) + self.bias
    

layer1 = LayerDense(4, 5)
layer2 = LayerDense(5, 2)

layer1.forward(X)
layer2.forward(layer1.output)
print(layer1.output)
print(layer2.output)


[[ 0.1152811   7.69522063  2.81115045  2.52504874  3.71647637]
 [ 0.47034874  6.11061917  2.53906165  1.50747988  2.3343289 ]
 [ 5.45075238  7.44991736  4.52664459 10.17779879  9.31098353]
 [ 0.84593703  3.7874768   1.90986991  1.38512263  1.81492571]]
[[18.55041775 -4.73602561]
 [13.49148266 -4.77259248]
 [17.53603842  3.22630427]
 [ 8.29456105 -2.62815743]]
